sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_converters/_ai_services.py

```diff
@@ -0,0 +1,243 @@
+import json
+
+from azure.ai.projects import AIProjectClient
+from azure.ai.projects.models import ThreadRun, RunStep, RunStepToolCallDetails, FunctionDefinition, ListSortOrder
+
+from typing import List
+
+# Constants.
+from ._models import _USER, _AGENT, _TOOL, _TOOL_CALL, _TOOL_CALLS, _FUNCTION
+
+# Message instances.
+from ._models import Message, SystemMessage, UserMessage, AssistantMessage, ToolCall
+
+# Intermediate definitions to hold results.
+from ._models import ToolDefinition, ConvertedResult
+
+# Utilities.
+from ._models import break_tool_call_into_messages, convert_message
+
+
+class AIAgentConverter:
+    """
+    A converter for AI agent data.
+
+    :param project_client: The AI project client used for API interactions.
+    :type project_client: AIProjectClient
+    """
+
+    def __init__(self, project_client: AIProjectClient):
+        self.project_client = project_client
+
+    def list_messages_chronological(self, thread_id: str):
+        """
+        Lists messages in chronological order for a given thread.
+
+        :param thread_id: The ID of the thread.
+        :type thread_id: str
+        :return: A list of messages in chronological order.
+        :rtype: List[Message]
+        """
+        messages = self.project_client.agents.list_messages(
+            thread_id=thread_id, limit=100, order=ListSortOrder.ASCENDING
+        )
+        return messages.data
+
+    def convert(self, thread_id: str, run_id: str) -> str:
+        """
+        Converts the agent run to a format suitable for the OpenAI API.
+
+        :param thread_id: The ID of the thread.
+        :type thread_id: str
+        :param run_id: The ID of the run.
+        :type run_id: str
+        :return: The converted data in dictionary format.
+        :rtype: str
+        """
+        # Make the API call once and reuse the result.
+        thread_run: ThreadRun = self.project_client.agents.get_run(thread_id=thread_id, run_id=run_id)
+
+        # We will collect messages in this accumulator.
+        final_messages: List[Message] = []
+
+        # Walk through the "user-facing" conversation history and start adding messages.
+        chronological_conversation = self.list_messages_chronological(thread_id)
+
+        # Each visible message in the conversation is a message from the user or the assistant, we collect
+        # both the text and timestamp, so we can recreate the chronological order.
+        in_my_current_run = False
+        for single_turn in chronological_conversation:
+            # Since this is the conversation of the entire thread and we are interested in a given run, we need to
+            # filter out the messages that came after the run.
+            if single_turn.run_id is not None:
+                if single_turn.run_id == run_id:
```

Comment:
List comprehension may be a better way to find all messages belong to a run id. See here as an example. https://www.w3schools.com/python/python_lists_comprehension.asp

Reply:
Agreed that list comprehension is neat, but notice that this is a greedy loop, where I want to add commentary on why each filtering is happening per a business and design decision reached. List comprehension due to its neatness would lose that non-trivial context. 

I prefer to commenting on top of a statement than in-line within the list comprehension.